server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java

```diff
@@ -0,0 +1,92 @@
+     *
+     * @return {@code Future<Boolean>} a future result expressing True/False decision
+     */
+    @Override
+    public Future<Boolean> isAuthorized(UserContext userContext, ResourceAction action, String resource) {
+        // parse the resource path based on the action and return the final resourceIDs for each resourceType
+        Map<ResourceType, String> resourceTypeToResourceId = getResourceIds(action, resource);
+        return Future.succeededFuture(resourceTypeToResourceId.entrySet().stream()
+                .anyMatch(entry -> isAuthorizedInternal(userContext.getSubject(), action, entry.getValue())));
+    }
+
+    private Map<ResourceType, String> getResourceIds(ResourceAction action, String resource) {
+        // TODO(aayush): need better way to do this?
+        Map<ResourceType, String> resultMap = new HashMap<>();
+        String[] segments = resource.split("/");
+        return resolve(ResourceType.ORG, action, List.of(segments), resultMap);
+    }
+
+    private Map<ResourceType, String> resolve(
+            ResourceType current, ResourceAction action, List<String> segments, Map<ResourceType, String> result
+    ) {
+        if (segments.isEmpty()) {
+            return result;
+        }
+        return switch (current) {
+            case ORG -> {
+                result.put(ResourceType.ORG, segments.remove(0));
+                yield resolve(ResourceType.TEAM, action, segments, result);
+            }
+            case TEAM -> {
+                result.put(ResourceType.TEAM, result.get(ResourceType.ORG) + ":" + segments.remove(0));
+                yield resolve(ResourceType.PROJECT, action, segments, result);
+            }
+            case PROJECT -> {
+                result.put(ResourceType.PROJECT, segments.remove(0));
+                yield resolve(action.getResourceType(), action, segments, result);
+            }
+            case TOPIC -> {
+                result.put(ResourceType.TOPIC, result.get(ResourceType.PROJECT) + ":" + segments.remove(0));
+                yield result;
+            }
+            case SUBSCRIPTION -> {
+                result.put(ResourceType.SUBSCRIPTION, result.get(ResourceType.PROJECT) + ":" + segments.remove(0));
+                yield result;
+            }
+        };
+    }
+
+    private List<String> getRolesForSubject(String subject, String resourceId) {
+        return configuration.getRoleBindings()
+                .getOrDefault(resourceId, new HashMap<>())
+                .getOrDefault(subject, new ArrayList<>());
+    }
+
+    private boolean isAuthorizedInternal(String subject, ResourceAction action, String resourceId) {
+        return getRolesForSubject(subject, resourceId).stream()
+                .anyMatch(role -> doesActionBelongToRole(role, action));
+    }
+
+    private boolean doesActionBelongToRole(String roleId, ResourceAction action) {
+        return configuration.getRoles().getOrDefault(roleId, new ArrayList<>()).contains(action);
```

Comment:
for debugging purpose, it would be good maintain (log) the information which node and which role returned true for authorised call.

Reply:
done